dvc/render/data.py

```diff
@@ -0,0 +1,211 @@
+from copy import deepcopy
+from functools import partial
+from typing import Dict, List, Optional, Set, Union
+
+from funcy import first, reraise
+
+from dvc.exceptions import DvcException
+from dvc.render.base import INDEX_FIELD, REVISION_FIELD
+
+
+class FieldsNotFoundError(DvcException):
+    def __init__(self, expected_fields, found_fields):
+        self.expected_fields = expected_fields
+        self.found_fields = found_fields
+
+        expected_str = ", ".join(expected_fields)
+        found_str = ", ".join(found_fields)
+        super().__init__(
+            f"Could not find all provided fields ('{expected_str}') "
+            f"in data fields ('{found_str}')."
+        )
+
+
+class PlotDataStructureError(DvcException):
+    def __init__(self):
+        super().__init__(
+            "Plot data extraction failed. Please see "
+            "https://man.dvc.org/plots for supported data formats."
+        )
+
+
+def _filter_fields(datapoints: List[Dict], fields: Set) -> List[Dict]:
+    if not fields:
+        return datapoints
+    assert isinstance(fields, set)
+
+    new_data = []
+    for data_point in datapoints:
+        keys = set(data_point.keys())
+        if keys & fields != fields:
+            raise FieldsNotFoundError(fields, keys)
+
+        new_data.append(
+            {key: value for key, value in data_point.items() if key in fields}
+        )
```

Comment:
May use `project()` from `funcy`:
```suggestion
        new_data.append(project(data_point, fields))
```